File: InsuraTech.Application/Common/Models/PagedResult.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace InsuraTech.Application.Common.Models
{
    public sealed class PagedResult<T>
    {
        public IEnumerable<T> Items { get; }
        public int TotalCount { get; }
        public int Page { get; }
        public int PageSize { get; }
        public int TotalPages => (int)Math.Ceiling(TotalCount / (double)PageSize);
        public bool HasNextPage => Page < TotalPages;
        public bool HasPreviousPage => Page > 1;
        public PagedResult(IEnumerable<T> items, int totalCount, int page, int pageSize)
        {
            Items = items;
            TotalCount = totalCount;
            Page = page;
            PageSize = pageSize;
        }


        public static PagedResult<T> Empty(int page, int pageSize) =>
                new(Enumerable.Empty<T>(), 0, page, pageSize);
    }
    
}

File: backend_v2/src/main/java/com/insuratech/application/common/models/PagedResult.java
package com.insuratech.application.common.models;

import java.util.List;

public record PagedResult<T>(
    List<T> items,
    int totalCount,
    int page,
    int pageSize
) {
    public int totalPages() {
        return (int) Math.ceil((double) totalCount / pageSize);
    }
    public boolean hasNextPage()     { return page < totalPages(); }
    public boolean hasPreviousPage() { return page > 1; }
}
